popcorn/apps/planner/__init__.py

```diff
@@ -52,114 +42,106 @@ def setup_defaults(self, loglevel=None, logfile=None, queue=None, strategy=None,
             previous_status = taste_soup(self.queue, self.app.conf['BROKER_URL'])
             time.sleep(HEARTBEAT_INTERVAL)
-            timestampe = int(round(time.time() * TIME_SCALE))
+            timestamp = int(round(time.time() * TIME_SCALE))
             status = taste_soup(self.queue, self.app.conf['BROKER_URL'])
             result = self.strategy.apply(
                 previous_status=previous_status,
-                previous_time=previous_timestampe,
+                previous_time=previous_timestamp,
                 status=status,
-                time=timestampe
+                time=timestamp
             )
+            if self.continue_flag != True:
+                info('[Planner] - [Stop Scan] ...')
+                break
             if result:
                 cmd = WorkerInstruction.generate_instruction_cmd(self.queue, result)
                 info('[Planner] - [Report New Demand] - %s' % str(cmd))
-                self.rpc_client.start('popcorn.apps.hub:hub_report_demand', type=InstructionType.WORKER, cmd=cmd)
-
-
-class Strategy(bootsteps.StartStopStep):
-
-    def __init__(self, p, **kwargs):
-        pass
-
-    def include_if(self, p):
-        return True
-
-    def create(self, p):
-        strategy_cls = STRATEGY_MAP[p.strategy_name]
-        p.strategy = instantiate(strategy_cls)
-        return self
-
-    def start(self, p):
-        pass
-
-    def stop(self, p):
-        print 'in stop'
-
-    def terminate(self, p):
-        print 'in terminate'
-
-
-class Loop(bootsteps.StartStopStep):
-    requires = (Strategy, RPCClient, )
-
-    def __init__(self, p, **kwargs):
-        pass
-
-    def include_if(self, p):
-        return True
-
-    def create(self, p):
-        self.process = Process(target=p.plan)
-        self.process.start()
-        return self
-
-    def start(self, p):
-        pass
-
-    def stop(self, p):
-        print 'in stop'
-
-    def terminate(self, p):
-        print 'in terminate'
-
-
-class Register(bootsteps.StartStopStep):
-    requires = (RPCClient, )
-
-    def __init__(self, p, **kwargs):
-        pass
-
-    def include_if(self, p):
-        return True
-
-    def create(self, p):
-        return self
-
-    def start(self, p):
-        p.rpc_client.start('popcorn.apps.planner:Planner', app=p.app, queue=p.queue, strategy_name=p.strategy_name)
-
-    def stop(self, p):
-        print 'in stop'
-
-    def terminate(self, p):
-        print 'in terminate'
+                hub_report_demand(type=InstructionType.WORKER, cmd=cmd)
+
+    def start(self, restart=True):
+        debug('[Planner] - [Start(restart=%s)] - %s' % (restart, self))
+        if self.get_thread():
+            if restart:
+                self.stop()
+                self.start(restart=False)
+            else:
+                raise Exception('thread already exist')
+        else:
+            self.continue_flag = True
+            Planner.threading_pool[self] = start_back_ground_task(self.plan)
```

Comment:
It's tricky to user class instance as key.

Need to define __hash__, return queue name.